feat: Editorial review, Relative color syntax (#32004)
mdn/content: files/en-us/web/css/color_value/hsl/index.md

slug: Web/CSS/color_value/hsl
page-type: css-function
browser-compat: css.types.color.hsl
spec-urls:
  - https://drafts.csswg.org/css-color-5/#relative-HSL
  - https://drafts.csswg.org/css-color/#the-hsl-notation
---

{{CSSRef}}

> The **`hsl()`** functional notation expresses an {{glossary("RGB", "sRGB")}} color according to its _hue_, _saturation_, and _lightness_ components. An optional _alpha_ component represents the color's transparency.

Author: I have updated all color function page intros to include a link to your fine new page ;-)

### Values

Below are descriptions of the allowed values for both absolute and [relative colors](/en-US/docs/Web/CSS/CSS_colors/Relative_colors).

> **Note:** Absolute and relative functional notation serializes to sRGB values, and the values of the red, green, blue components may be rounded in serialization.

#### Absolute value syntax

```text
hsl(H S L[ / A])
```

The parameters are as follows:

- `H`
  - : A {{CSSXref("&lt;number&gt;")}}, an {{CSSXref("&lt;angle&gt;")}}, or the keyword `none` (equivalent to `0deg` in this case). This value represents the hue angle. More details on this type can be found on the {{CSSXref("&lt;hue&gt;")}} reference page.
- `S`
  - : A {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). This value represents the color's saturation. Here `100%` is completely saturated, while `0%` is completely unsaturated (gray).
- `L`
  - : A {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). This value represents the color's lightness. Here `100%` is white, `0%` is black, and `50%` is "normal".
- `A` {{optional_inline}}

>   - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` explicitly specifies no alpha channel. If the `alpha` channel value is not explicitly specified, it defaults to 100%. If included, the value is preceded by a slash (`/`).

Reviewer: The `alpha` here should not be code, as you can't use "alpha" in the absolute syntax. Move the sentence to the relative alpha. Here's a suggestion for this section:
```suggestion
  - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` explicitly specifies no alpha channel. If the alphatransparency channel value is not explicitly specified, it defaults to 100%. If included, the value is preceded by a slash (`/`).
```

Author: I've just changed `alpha` to `A`, which is what we do in most of the other pages.

#### Relative value syntax

```text
hsl(from <color> H S L[ / A])
```

The parameters are as follows:

- `from <color>`
  - : The keyword `from` is always included when defining a relative color, followed by a {{cssxref("&lt;color&gt;")}} value representing the **origin color**. This is the original color that the relative color is based on. Note that the origin color can be _any_ valid {{cssxref("&lt;color&gt;")}} syntax, including another relative color.
- `H`
  - : A {{CSSXref("&lt;number&gt;")}}, an {{CSSXref("&lt;angle&gt;")}}, or the keyword `none` (equivalent to `0deg` in this case). This represents the hue angle of the output color.
- `S`
  - : A {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). This represents the saturation of the output color. Here `100%` is completely saturated, while `0%` is completely unsaturated (gray).
- `L`
  - : A {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). This represents the lightness of the output color. Here `100%` is white, `0%` is black, and `50%` is "normal".
- `A` {{optional_inline}}

>   - : An {{CSSXref("&lt;alpha-value&gt;")}} where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` explicitly specifies no alpha channel. This represents the alpha channel value of the output color. If the `A` channel value is not explicitly specified, it defaults to the alpha channel value of the origin color. If included, the value is preceded by a slash (`/`).

Reviewer: ```suggestion
  - :  An {{CSSXref("&lt;alpha-value&gt;")}} representing the `alpha` channel value of the output color, where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` explicitly specifies no alpha channel. If the `alpha` value is not explicitly specified, it defaults to the alpha channel value of the origin color. If included, the value is preceded by a slash (`/`).
```
changed "If the `A` channel value" to "If the `alpha` value" to demonstrate that `alpha` is the channel name. (it's on line 82, but good to have here, and removed from absolute definition on line 57)

Author: OK, I've made the first change as I thought it was useful to provide an explicit definition, but I've not used `alpha`; I've just used alpha.

And I've not changed `A` to `alpha`.

The thing is, the functional notation we've used on these pages to signify the alpha channel is `A`. `alpha` is a different thing — this is the variable made available inside a relative color function that contains the alpha channel value of the origin color. You can use this to define the `A` channel value of the output color if you want to (you could define it as `alpha`, or as something more complex like `calc(alpha + 0.1)`), or you could choose to not use `alpha`, and instead define the output color's `A` channel value as an absolute value like `0.5` or `0.25`.

#### Defining relative color output channel components

> When using relative color syntax inside an `hsl()` function, the browser converts the origin color into an equivalent HSL color (if it is not already specified as such) The color is defined as three distinct color channel values — `h` (hue), `s` (saturation), and `l` (lightness) — plus an alpha channel value (`alpha`). These channel values are made available inside the function to be used when defining the output color channel values:

Reviewer: ```suggestion
When using relative color syntax inside an `hsl()` function, the browser converts the origin color to {{glossary("RGB", "sRGB")}}, if necessary, then into an equivalent HSL color (if it is not already specified as such) The color is defined as three distinct color channel values — `h` (hue), `s` (saturation), and `l` (lightness) — plus an alpha channel value (`alpha`). These channel values are made available inside the function to be used when defining the output color channel values:
```
not sure if this edit is beyond what we need to convey to our readers.

Reviewer: you say it on line 98, so maybe not needed.

Author: Yeah, not needed. I don't think this is correct, anyway — at this stage, the color is still being represented as the HSL equivalent. It is only at the output stage where it is converted into an sRGB equivalent.

> - The `h` value is resolved to a {{cssxref("&lt;number&gt;")}} between `0` and `360` that represents the origin color's {{cssxref("&lt;hue&gt;")}} degree value.

Reviewer: ```suggestion
- The `h` value is resolved to a {{cssxref("&lt;number&gt;")}} between `0` and `360`, inclusive, that represents the origin color's {{cssxref("&lt;hue&gt;")}} degree value.
```

Author: updated

> - The `s` and `l` values are each resolved to a `<number>` between `0` and `100`.

Reviewer: ```suggestion
- The `s` and `l` values are each resolved to a `<number>` between `0` and `100`, inclusive, where `100` is equivalent to `100%`.
```

Author: updated

- The `alpha` value is resolved to a `<number>` between `0` and `1`, inclusive.

When defining a relative color, the different channels of the output color can be expressed in several different ways. Below, we'll study some examples to illustrate these.

> It is important to bear in mind that, in the first two examples below, we are using relative color syntax. However, the first one outputs the same color as the origin color and the second one outputs a color not based on the origin color at all. They don't really create relative colors! You'd be unlikely to ever use these in a real codebase, and would probably just use an absolute color value instead. We included these examples as a starting point for learning about relative `hsl()` syntax.

Reviewer: ```suggestion
In the first two examples below, we are using relative color syntax. However, the first one outputs the same color as the origin color and the second one outputs a color not based on the origin color at all. They don't really create relative colors! You'd be unlikely to ever use these in a real codebase, and would probably just use an absolute color value instead. We included these examples as a starting point for learning about relative `hsl()` syntax.
```

Author: updated

Let's start with an origin color of `rgb(255 0 0)` (equivalent to `red`). The following function outputs the same color as the origin color — it uses the origin color's `h`, `s`, and `l` channel values as the output channel values:

```css

> hsl(from rgb(255 0 0) h s l)

Reviewer: ```suggestion
hsl(from rgb(255 0 0) h s l)
/* Output color: hsl(0 100% 50%) */
```

Author: It's not though; the output color is `color(srgb 1 0 0)`. I've handled this in a slightly different way, and put an equivalent on all the other pages.

```

> > **Note:** As mentioned above, if the output color is using a different color model to the origin color, the origin color is converted to the same model as the output color in the background so that it can be represented in a way that is compatible (i.e. using the same channels). For example, in the above case the {{cssxref("color_value/rgb", "rgb()")}} color `rgb(255 0 0)` is converted to an sRGB `color()` function equivalent to `hsl(0 100 50)`: `color(srgb 1 0 0)`.

Reviewer: ```suggestion
> **Note:** As mentioned above, if the output color is using a different color model to the origin color, the origin color is converted to the same model as the output color in the background so that it can be represented in a way that is compatible (i.e. using the same channels). For example, in the above case the {{cssxref("color_value/rgb", "rgb()")}} color `rgb(255 0 0)` is converted to an sRGB `color()` function equivalent to `hsl(0 100% 50%)`: `color(srgb 1 0 0)`.
```
not sure if my edit is correct, but based on line 106, it is.

This function uses absolute values for the output color's channel values, outputting a completely different color not based on the origin color:

```css
hsl(from rgb(255 0 0) 240 60% 70%)
```

In the above case, the output color is the sRGB `color()` equivalent of `hsl(240 60% 70%)`: `color(srgb 0.52 0.52 0.88)`.

The following function creates a relative color based on the origin color:

```css
hsl(from rgb(255 0 0) h 30% 60%)
```

This example:

- Converts the `rgb()` origin color to an equivalent of the `hsl()` representation in the sRGB color space — `color(srgb 1 0 0)` behind the scenes.
- Sets the `H` channel value for the output color to that of the origin `color()` equivalent's `H` value — `1`.
- Sets the output color's `S` and `L` channel values to new values not based on the origin color: `30%` and `60%`, respectively.

The above example's output color is `color(srgb 0.72 0.48 0.48)`.

> In the examples we've seen so far in this section, the alpha channels have not been explicitly specified, for either the origin or output colors. When the output color alpha channel is not specified, it defaults to the same value as the origin color alpha channel. When the origin color alpha channel is not specified (and it is not a relative color), it defaults to `1`. Therefore, the origin and output channel values are `1` for the above examples.

Reviewer: ```suggestion
> **Note:** In the examples we've seen so far in this section, the alpha channels have not been explicitly specified for either the origin or output colors. When the output color alpha channel is not specified, it defaults to the same value as the origin color alpha channel. When the origin color alpha channel is not specified (and it is not a relative color), it defaults to `1`. Therefore, the origin and output channel values are `1` for the above examples.
```

Author: I don't see this as a note; it is directly related to the alpha examples that came after it, whereas a note makes it feel like more of an aside. No change made.

Let's look at some examples that specify origin and output alpha channel values. The first one specifies the output alpha channel value as being the same as the origin alpha channel value, whereas the second one specifies a different output alpha channel value, unrelated to the origin alpha channel value.

```css
hsl(from rgb(255 0 0 / 0.8) h s l / alpha)

> /* Computed output color: color(srgb 1 0 0 / 0.8) */

Reviewer: isn't the output color an HSL color?:
```suggestion
/* Computed output color: color(srgb 1 0 0 / 0.8) */
/* Output color: hsl(0 100% 50% / 0.8) */
```

Author: I don't think this makes sense. No change made.

hsl(from rgb(255 0 0 / 0.8) h s l / 0.5)
/* Computed output color: color(srgb 1 0 0 / 0.5) */
```

> In the following example, the `rgb()` origin color is again converted into an equivalent of the `hwb()` representation in the sRGB color space — `color(srgb 1 0 0 / 0.8)` in this case. {{cssxref("calc")}} calculations are applied to the `H`, `S`, `L`, and `A` values, resulting in an output color of `color(srgb 171 171 -190 / 0.7)`:

Reviewer: ```suggestion
In the following example, the `rgb()` origin color is again converted into an equivalent of the `hsl()` representation in the sRGB color space — `color(srgb 1 0 0 / 0.8)` in this case. {{cssxref("calc")}} calculations are applied to the `h`, `s`, `l`, and `alpha` values, resulting in an output color of `color(srgb 171 171 -190 / 0.7)`:
```

Reviewer: This doesn't seem right. srgb colors go from 0 to 1. Here we have 171 and -190.

Author: Ooops, yes, this is because of the bugs that still exist in Chrome around the calc values in HSL colors. I've corrected it to the right value — `color(srgb 0.72 0.72 0.08 / 0.7)`

```css

> hsl(from rgb(255 0 0 / 0.8) calc(h + 60) calc(s - 20) calc(l - 10) / calc(alpha - 0.1))

Reviewer: ```suggestion
hsl(from rgb(255 0 0 / 0.8) calc(h + 60) calc(s - 20) calc(l - 10) / calc(alpha - 0.1))
/* output color: hsl(60 80% 40% / 0.7) */
```

Author: Again, I don't think this makes sense

## Syntax

```css
/* Absolute values */
hsl(120deg 75% 25%)
hsl(120 75% 25%) /* deg unit is optional */
hsl(120deg 75% 25% / 60%)
hsl(none 75% 25%)

/* Relative values */
hsl(from green h s l / 0.5)
hsl(from #0000FF h s calc(l + 20))
hsl(from rgb(200 0 0) calc(h + 30) s calc(l + 30))
```

The `hsla()` function can also be used to express sRGB colors. This is an alias for `hsl()` that accepts the same parameters.

> **Note:** The computed values of colors expressed using `hsl()` syntax are serialized to {{CSSXref("color_value/rgb", "rgb()")}} values in the case of absolute colors, and {{CSSXref("color_value/color", "color(srgb ... )")}} values in the case of relative colors. The values of the red, green, blue components may be rounded in serialization.

> > **Note:** `hsl()`/`hsla()` can also be written in a legacy form in which all values are separated with commas, for example `hsl(120deg, 75%, 25%)`. Mixing number and percent value types is not valid in the comma-separated legacy syntax (i.e. the `S` and `L` values must be either all numbers or all percentages), and the `none` value is also not permitted.

Author: Makes sense. I've just done a bit of testing, and it looks like the % unit is mandatory when using the legacy commas. As a result, I've updated the note to:

> **Note:** `hsl()`/`hsla()` can also be written in a legacy form in which all values are separated with commas, for example `hsl(120deg, 75%, 25%)`. The `none` value is not permitted in the comma-separated legacy syntax, and the `%` units are required.

### Values

Below are descriptions of the allowed values for both absolute and [relative colors](/en-US/docs/Web/CSS/CSS_colors/Relative_colors).

#### Absolute value syntax

```text
hsl(H S L[ / A])
```

The parameters are as follows:

- `H`

>   - : A {{CSSXref("&lt;number&gt;")}}, an {{CSSXref("&lt;angle&gt;")}}, or the keyword `none` (equivalent to `0deg` in this case) representing the color's {{CSSXref("&lt;hue&gt;")}} angle.

Author: I don't think a mention is needed here.

#### Relative value syntax

```text
hsl(from <color> H S L[ / A])
```

The parameters are as follows:

- `from <color>`
  - : The keyword `from` is always included when defining a relative color, followed by a {{cssxref("&lt;color&gt;")}} value representing the **origin color**. This is the original color that the relative color is based on. The origin color can be _any_ valid {{cssxref("&lt;color&gt;")}} syntax, including another relative color.
- `H`
  - : A {{CSSXref("&lt;number&gt;")}}, an {{CSSXref("&lt;angle&gt;")}}, or the keyword `none` (equivalent to `0deg` in this case) representing the output color's {{CSSXref("&lt;hue&gt;")}} angle.
- `S`
  - : A {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). This represents the saturation of the output color. Here `100%` is completely saturated, while `0%` is completely unsaturated (gray).
- `L`
  - : A {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). This represents the lightness of the output color. Here `100%` is white, `0%` is black, and `50%` is "normal".
- `A` {{optional_inline}}

>   - : An {{CSSXref("&lt;alpha-value&gt;")}} representing the alpha channel value of the output color, where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` can be used to explicitly specify no alpha channel. This represents the alpha channel value of the output color. If the `A` channel value is not explicitly specified, it defaults to the alpha channel value of the origin color. If included, the value is preceded by a slash (`/`).

Reviewer: ```suggestion
  - : An {{CSSXref("&lt;alpha-value&gt;")}} representing the alpha channel value of the output color, where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (full opacity), or the keyword `none` which can be used to explicitly specify no alpha channel. This represents the alpha channel value of the output color. If the `A` channel value is not explicitly specified, it defaults to the alpha channel value of the origin color. If included, the value is preceded by a slash (`/`).
```

Author: The original was more correct than your suggestion, IMO. However, we've been chewing on this one quite a bit in this review cycle. I've elected to break the sentence into two. I've also removed a sentence later on that repeats what is said at the start of the para; and I've updated "full opacity" to "fully opaque", to be consistent with "fully transparent". I ended up with this:

- : An {{CSSXref("&lt;alpha-value&gt;")}} representing the alpha channel value of the output color, where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (fully opaque). Additionally, the keyword `none` can be used to explicitly specify no alpha channel. If the `A` channel value is not explicitly specified, it defaults to the alpha channel value of the origin color. If included, the value is preceded by a slash (`/`).

> > **Note:** The `hsla()` alias can also be used to output relative colors, and to specify origin colors. When using `hsla()` to output a relative color, you must use the comma-less modern syntax and can't mix percentages and numbers (i.e. for the `S` and `L` values).

Reviewer: ```suggestion
> **Note:** The `hsla()` alias can also be used to output relative colors, and to specify origin colors. When using `hsla()` to output a relative color, you must use the comma-less modern syntax.
```

Author: updated

> 

Reviewer: ```suggestion

> **Note:** To fully enable the representation of the full spectrum of visible colors, the output of relative `hsl()` color functions are serialized to `color(srgb)`. That means that querying the output color value via the {{DOMxRef("HTMLElement.style")}} property or the {{DOMxRef("CSSStyleDeclaration.getPropertyValue()")}} method returns the output color as a [`color(srgb ...)`](/en-US/docs/Web/CSS/color_value/color) value.
```

Reviewer: or on line 103

Author: added.

#### Defining relative color output channel components

When using relative color syntax inside an `hsl()` function, the browser converts the origin color into an equivalent HSL color (if it is not already specified as such). The color is defined as three distinct color channel values — `h` (hue), `s` (saturation), and `l` (lightness) — plus an alpha channel value (`alpha`). These channel values are made available inside the function to be used when defining the output color channel values:

- The `h` value is resolved to a {{cssxref("&lt;number&gt;")}} between `0` and `360`, inclusive, that represents the origin color's {{cssxref("&lt;hue&gt;")}} degree value.

> - The `s` and `l` values are each resolved to a `<number>` between `0` and `100`, inclusive, where `100` is equivalent to `100%`.

Reviewer: Do we want to add a comment or note that "while these values must be percents when defining HSL, when creating relative colors, all values are resolved to numbers"?

Author: No; that's not true anyway - in non-legacy syntax at least, the % signs are optional. I tested this recently.

Let's start with an origin color of `rgb(255 0 0)` (equivalent to `hsl(0 100% 50%)`). The following function outputs the same color as the origin color — it uses the origin color's `h`, `s`, and `l` channel values as the output channel values:

```css
hsl(from rgb(255 0 0) h s l)
```

This function's output color is the sRGB `color()` equivalent of `hsl(0, 100%, 50%)`: `color(srgb 1 0 0)`.

The next function uses absolute values for the output color's channel values, outputting a completely different color not based on the origin color:

```css
hsl(from rgb(255 0 0) 240 60% 70%)
```

In the above case, the output color is the sRGB `color()` equivalent of `hsl(240 60% 70%)`: `color(srgb 0.52 0.52 0.88)`.

The following function creates a relative color based on the origin color:

```css
hsl(from rgb(255 0 0) h 30% 60%)
```

This example:

- Converts the origin color (`rgb(255 0 0)`) into an `hsl()` equivalent (`hsl(0 100% 50%)`).
- Sets the `H` channel value for the output color to those of the origin color `hsl()` equivalent's `H` channel value — `0`.

> - Sets the output color's `S` and `L` channel values to new values not based on the origin color: `30%` and `60%`, respectively. The output `hsl()` color is `hsl(0 30% 60%)`.
> 
> The final output color is the equivalent color in the sRGB color space — `color(srgb 0.72 0.48 0.48)`.

Reviewer: ```suggestion
- Sets the output color's `S` and `L` channel values to new values not based on the origin color: `30%` and `60%`, respectively. The output `hsl()` color is equivalent to `hsl(0 30% 60%)` in the sRGB color space — `color(srgb 0.72 0.48 0.48)`.
```

Author: I think putting these two together makes it slightly more confusing, as the final output information is not really part of this step. To hopefully improve things, I've updated the second line to:

> The final output color is the equivalent of `hsl(0 30% 60%)` in the sRGB color space — `color(srgb 0.72 0.48 0.48)`.

> **Note:** As mentioned above, if the output color is using a different color model to the origin color, the origin color is converted to the same model as the output color in the background so that it can be represented in a way that is compatible (i.e. using the same channels).

> In the examples we've seen so far in this section, the alpha channels have not been explicitly specified, for either the origin or output colors. When the output color alpha channel is not specified, it defaults to the same value as the origin color alpha channel. When the origin color alpha channel is not specified (and it is not a relative color), it defaults to `1`. Therefore, the origin and output channel values are `1` for the above examples.

Reviewer: ```suggestion
In the examples we've seen so far in this section, the alpha channels have not been explicitly specified, for either the origin or output colors. When the output color alpha channel is not specified, it defaults to the same value as the origin color alpha channel. When the origin color alpha channel is not specified (and it is not a relative color), it defaults to `1`. Therefore, the origin and output alpha channel values are `1` for the above examples.
```

Author: Updated, across all color function docs.

Let's look at some examples that specify origin and output alpha channel values. The first one specifies the output alpha channel value as being the same as the origin alpha channel value, whereas the second one specifies a different output alpha channel value, unrelated to the origin alpha channel value.

```css
hsl(from rgb(255 0 0 / 0.8) h s l / alpha)
/* Computed output color: color(srgb 1 0 0 / 0.8) */

hsl(from rgb(255 0 0 / 0.8) h s l / 0.5)
/* Computed output color: color(srgb 1 0 0 / 0.5) */
```

> In the following example, the `rgb()` origin color is again converted into an `hsl()` representation — `hsl(0 100% 50% / 0.8)`. {{cssxref("calc")}} calculations are applied to the `H`, `S`, `L`, and `A` values, and the final output color is `color(srgb 0.72 0.72 0.08 / 0.7)`:

Reviewer: ```suggestion
In the following example, the `rgb()` origin color is again converted into an `hsl()` representation — `hsl(0 100% 50% / 0.8)`. {{cssxref("calc")}} calculations are applied to the `H`, `S`, `L`, and `A` values, and the final output color is equivalent to `hsl(60 80%30% / 0.7)` : `color(srgb 0.72 0.72 0.08 / 0.7)`:
```

Author: updated

- `S`
  - : A {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). This value represents the color's saturation. Here `100%` is completely saturated, while `0%` is completely unsaturated (gray).
- `L`
  - : A {{CSSXref("&lt;percentage&gt;")}} or the keyword `none` (equivalent to `0%` in this case). This value represents the color's lightness. Here `100%` is white, `0%` is black, and `50%` is "normal".
- `A` {{optional_inline}}
  - : An {{CSSXref("&lt;alpha-value&gt;")}} representing the alpha channel value of the color, where the number `0` corresponds to `0%` (fully transparent) and `1` corresponds to `100%` (fully opaque). Additionally, the keyword `none` can be used to explicitly specify no alpha channel. If the `A` channel value is not explicitly specified, it defaults to 100%. If included, the value is preceded by a slash (`/`).

> **Note:** See [Missing color components](/en-US/docs/Web/CSS/color_value#missing_color_components) for more information on the effect of `none`.

> > **Note:** Absolute `hsl()` colors are serialized to {{CSSXref("color_value/rgb", "rgb()")}} values. The values of the red, green, and blue components may be rounded in serialization.

Reviewer: Create a glossary page for "serialized" to explain what serializing means, especially as it relates to colors.

Link to this new page from:
- [ ]  hsl()
- [ ] rgb(),
- [ ] hwb()
- [ ] color space glossary.
- [ ] css colors module landing page
- [ ] using relative colors page

Reviewer: Created an issue

- The `s` and `l` values are each resolved to a `<number>` between `0` and `100`, inclusive, where `100` is equivalent to `100%`.
- The `alpha` value is resolved to a `<number>` between `0` and `1`, inclusive.

When defining a relative color, the different channels of the output color can be expressed in several different ways. Below, we'll study some examples to illustrate these.

In the first two examples below, we are using relative color syntax. However, the first one outputs the same color as the origin color and the second one outputs a color not based on the origin color at all. They don't really create relative colors! You'd be unlikely to ever use these in a real codebase, and would probably just use an absolute color value instead. We included these examples as a starting point for learning about relative `hsl()` syntax.

Let's start with an origin color of `rgb(255 0 0)` (equivalent to `hsl(0 100% 50%)`). The following function outputs the same color as the origin color — it uses the origin color's `h`, `s`, and `l` channel values as the output channel values:

```css
hsl(from rgb(255 0 0) h s l)
```

> This function's output color is the sRGB `color()` equivalent of `hsl(0, 100%, 50%)`: `color(srgb 1 0 0)`.

Author: good spot; fixed!

```css
hsl(from rgb(255 0 0) 240 60% 70%)
```

In the above case, the output color is the sRGB `color()` equivalent of `hsl(240 60% 70%)`: `color(srgb 0.52 0.52 0.88)`.

The following function creates a relative color based on the origin color:

```css
hsl(from rgb(255 0 0) h 30% 60%)
```

This example:

- Converts the origin color (`rgb(255 0 0)`) into an `hsl()` equivalent (`hsl(0 100% 50%)`).
- Sets the `H` channel value for the output color to those of the origin color `hsl()` equivalent's `H` channel value — `0`.
- Sets the output color's `S` and `L` channel values to new values not based on the origin color: `30%` and `60%`, respectively. The output `hsl()` color is `hsl(0 30% 60%)`.

The final output color is the equivalent of `hsl(0 30% 60%)` in the sRGB color space — `color(srgb 0.72 0.48 0.48)`.

> **Note:** As mentioned above, if the output color is using a different color model to the origin color, the origin color is converted to the same model as the output color in the background so that it can be represented in a way that is compatible (i.e. using the same channels).

> In the examples we've seen so far in this section, the alpha channels have not been explicitly specified, for either the origin or output colors. When the output color alpha channel is not specified, it defaults to the same value as the origin color alpha channel. When the origin color alpha channel is not specified (and it is not a relative color), it defaults to `1`. Therefore, the origin and output alpha channel values are `1` for the above examples.

Author: updated, on this and all the other color function pages.
